The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5001759371/jobs/8960856429) [(plain)](https://github.com/rust-lang/rust/commit/f100c3c1fbdedf0f5a3b45a514d593f311778384/checks/13545576262/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [ui] tests/ui/offset-of/offset-of-dst-field.rs stdout ----

error: Error: expected failure status (Some(1)) but received status Some(101).
status: exit status: 101
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/offset-of/offset-of-dst-field.rs" "-Zthreads=1" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/checkout/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/offset-of/offset-of-dst-field" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/offset-of/offset-of-dst-field/auxiliary"
stdout: none
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
  --> fake-test-src-base/offset-of/offset-of-dst-field.rs:36:5
   |
   |
LL |     offset_of!(Alpha, z); //~ ERROR the size for values of type
   |
   = help: the trait `Sized` is not implemented for `[u8]`
   = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)


error[E0277]: the size for values of type `(dyn Trait + 'static)` cannot be known at compilation time
  --> fake-test-src-base/offset-of/offset-of-dst-field.rs:37:5
   |
LL |     offset_of!(Beta, z); //~ ERROR the size for values of type
   |
   |
   = help: the trait `Sized` is not implemented for `(dyn Trait + 'static)`

error[E0277]: the size for values of type `Extern` cannot be known at compilation time
  --> fake-test-src-base/offset-of/offset-of-dst-field.rs:38:5
   |
   |
LL |     offset_of!(Gamma, z); //~ ERROR the size for values of type
   |
   = help: the trait `Sized` is not implemented for `Extern`
   = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)


error[E0277]: the size for values of type `dyn Trait` cannot be known at compilation time
  --> fake-test-src-base/offset-of/offset-of-dst-field.rs:44:5
   |
LL |     offset_of!(Delta<dyn Trait>, z); //~ ERROR the size for values of type
   |
   = help: the trait `Sized` is not implemented for `dyn Trait`
   = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)


error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
  --> fake-test-src-base/offset-of/offset-of-dst-field.rs:43:5
   |
LL |     offset_of!(Delta<Alpha>, z); //~ ERROR the size for values of type
   |
   |
   = help: within `Alpha`, the trait `Sized` is not implemented for `[u8]`
note: required because it appears within the type `Alpha`
  --> fake-test-src-base/offset-of/offset-of-dst-field.rs:5:8
LL | struct Alpha {
   |        ^^^^^
   = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)


error: internal compiler error: compiler/rustc_hir_typeck/src/writeback.rs:697:17: writeback: `Delta<dyn Trait>` has inference variables
  --> fake-test-src-base/offset-of/offset-of-dst-field.rs:44:5
   |
LL |     offset_of!(Delta<dyn Trait>, z); //~ ERROR the size for values of type
   |
   = note: this error: internal compiler error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)

thread 'rustc' panicked at 'Box<dyn Any>', /checkout/compiler/rustc_errors/src/lib.rs:994:33
thread 'rustc' panicked at 'Box<dyn Any>', /checkout/compiler/rustc_errors/src/lib.rs:994:33
stack backtrace:
   0:     0x7fb5735d9494 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hd2d8e2309d1fed43
   1:     0x7fb573640148 - core::fmt::write::h6d7af34a838da8d4
   2:     0x7fb5735cdba1 - std::io::Write::write_fmt::h9a034920726890be
   3:     0x7fb5735d92a1 - std::sys_common::backtrace::print::h4adf185dae0f6169
   4:     0x7fb5735dc41a - std::panicking::default_hook::{{closure}}::ha2872a2590c8ebf3
   5:     0x7fb5735dc0fc - std::panicking::default_hook::h981bc07c11c4931b
   6:     0x7fb57408922b - rustc_driver_impl[40b6097c0ea8d6ac]::install_ice_hook::{closure#0}
   7:     0x7fb5735dcb27 - std::panicking::rust_panic_with_hook::he9f338882ab0bc02
   8:     0x7fb574681a73 - std[88ff7485e6789050]::panicking::begin_panic::<rustc_errors[6c11ecac5352b63e]::ExplicitBug>::{closure#0}
   9:     0x7fb574671016 - std[88ff7485e6789050]::sys_common::backtrace::__rust_end_short_backtrace::<std[88ff7485e6789050]::panicking::begin_panic<rustc_errors[6c11ecac5352b63e]::ExplicitBug>::{closure#0}, !>
  10:     0x7fb573e491b6 - std[88ff7485e6789050]::panicking::begin_panic::<rustc_errors[6c11ecac5352b63e]::ExplicitBug>
  11:     0x7fb574731461 - <rustc_errors[6c11ecac5352b63e]::HandlerInner>::span_bug::<rustc_span[150ea94e489d3a47]::span_encoding::Span, alloc[9141ddfa543d6991]::string::String>
  12:     0x7fb574731140 - <rustc_errors[6c11ecac5352b63e]::Handler>::span_bug::<rustc_span[150ea94e489d3a47]::span_encoding::Span, alloc[9141ddfa543d6991]::string::String>
  13:     0x7fb5746998b8 - rustc_middle[faf83cb2799b7ed6]::util::bug::opt_span_bug_fmt::<rustc_span[150ea94e489d3a47]::span_encoding::Span>::{closure#0}
  14:     0x7fb57469990c - rustc_middle[faf83cb2799b7ed6]::ty::context::tls::with_opt::<rustc_middle[faf83cb2799b7ed6]::util::bug::opt_span_bug_fmt<rustc_span[150ea94e489d3a47]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  15:     0x7fb574692ee4 - rustc_middle[faf83cb2799b7ed6]::ty::context::tls::with_context_opt::<rustc_middle[faf83cb2799b7ed6]::ty::context::tls::with_opt<rustc_middle[faf83cb2799b7ed6]::util::bug::opt_span_bug_fmt<rustc_span[150ea94e489d3a47]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  16:     0x7fb573e42a87 - rustc_middle[faf83cb2799b7ed6]::util::bug::span_bug_fmt::<rustc_span[150ea94e489d3a47]::span_encoding::Span>
  17:     0x7fb574642ec7 - <rustc_hir_typeck[5c35d48b0356938b]::writeback::WritebackCx>::visit_offset_of_container_types
  18:     0x7fb5745cd8bc - <rustc_hir_typeck[5c35d48b0356938b]::fn_ctxt::FnCtxt>::resolve_type_vars_in_body
  19:     0x7fb574665918 - rustc_hir_typeck[5c35d48b0356938b]::typeck
  20:     0x7fb575d56548 - rustc_query_system[ef73c22083ab44c7]::query::plumbing::try_execute_query::<rustc_query_impl[62a88670161651c2]::DynamicConfig<rustc_query_system[ef73c22083ab44c7]::query::caches::VecCache<rustc_span[150ea94e489d3a47]::def_id::LocalDefId, rustc_middle[faf83cb2799b7ed6]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[62a88670161651c2]::plumbing::QueryCtxt, false>
  21:     0x7fb575b474ca - rustc_query_impl[62a88670161651c2]::get_query_non_incr::typeck
  22:     0x7fb57464f039 - rustc_middle[faf83cb2799b7ed6]::ty::query::query_get_at::<rustc_query_system[ef73c22083ab44c7]::query::caches::VecCache<rustc_span[150ea94e489d3a47]::def_id::LocalDefId, rustc_middle[faf83cb2799b7ed6]::query::erase::Erased<[u8; 8usize]>>>
  23:     0x7fb574664bc2 - rustc_hir_typeck[5c35d48b0356938b]::used_trait_imports
  24:     0x7fb575d56548 - rustc_query_system[ef73c22083ab44c7]::query::plumbing::try_execute_query::<rustc_query_impl[62a88670161651c2]::DynamicConfig<rustc_query_system[ef73c22083ab44c7]::query::caches::VecCache<rustc_span[150ea94e489d3a47]::def_id::LocalDefId, rustc_middle[faf83cb2799b7ed6]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[62a88670161651c2]::plumbing::QueryCtxt, false>
  25:     0x7fb575b481ea - rustc_query_impl[62a88670161651c2]::get_query_non_incr::used_trait_imports
  26:     0x7fb574950de9 - rustc_middle[faf83cb2799b7ed6]::ty::query::query_get_at::<rustc_query_system[ef73c22083ab44c7]::query::caches::VecCache<rustc_span[150ea94e489d3a47]::def_id::LocalDefId, rustc_middle[faf83cb2799b7ed6]::query::erase::Erased<[u8; 8usize]>>>
  27:     0x7fb57496730f - rustc_hir_analysis[2d80267229b5d511]::check_crate
  28:     0x7fb57417d369 - rustc_interface[6d78e37740b0e5ff]::passes::analysis
  29:     0x7fb575e2aea7 - <rustc_query_impl[62a88670161651c2]::dynamic_query::analysis::{closure#2} as core[85c62f5f18d12f5e]::ops::function::FnOnce<(rustc_middle[faf83cb2799b7ed6]::ty::context::TyCtxt, ())>>::call_once
  30:     0x7fb575cc82a2 - rustc_query_system[ef73c22083ab44c7]::query::plumbing::try_execute_query::<rustc_query_impl[62a88670161651c2]::DynamicConfig<rustc_query_system[ef73c22083ab44c7]::query::caches::SingleCache<rustc_middle[faf83cb2799b7ed6]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[62a88670161651c2]::plumbing::QueryCtxt, false>
  31:     0x7fb575b237a6 - rustc_query_impl[62a88670161651c2]::get_query_non_incr::analysis
  32:     0x7fb5740a57ad - <rustc_middle[faf83cb2799b7ed6]::ty::context::GlobalCtxt>::enter::<rustc_driver_impl[40b6097c0ea8d6ac]::run_compiler::{closure#1}::{closure#2}::{closure#4}, core[85c62f5f18d12f5e]::result::Result<(), rustc_span[150ea94e489d3a47]::ErrorGuaranteed>>
  33:     0x7fb5740db888 - <rustc_interface[6d78e37740b0e5ff]::interface::Compiler>::enter::<rustc_driver_impl[40b6097c0ea8d6ac]::run_compiler::{closure#1}::{closure#2}, core[85c62f5f18d12f5e]::result::Result<core[85c62f5f18d12f5e]::option::Option<rustc_interface[6d78e37740b0e5ff]::queries::Linker>, rustc_span[150ea94e489d3a47]::ErrorGuaranteed>>
  34:     0x7fb5740a4400 - rustc_span[150ea94e489d3a47]::set_source_map::<core[85c62f5f18d12f5e]::result::Result<(), rustc_span[150ea94e489d3a47]::ErrorGuaranteed>, rustc_interface[6d78e37740b0e5ff]::interface::run_compiler<core[85c62f5f18d12f5e]::result::Result<(), rustc_span[150ea94e489d3a47]::ErrorGuaranteed>, rustc_driver_impl[40b6097c0ea8d6ac]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  35:     0x7fb574096ae9 - <scoped_tls[ea298804ec9fbff7]::ScopedKey<rustc_span[150ea94e489d3a47]::SessionGlobals>>::set::<rustc_interface[6d78e37740b0e5ff]::interface::run_compiler<core[85c62f5f18d12f5e]::result::Result<(), rustc_span[150ea94e489d3a47]::ErrorGuaranteed>, rustc_driver_impl[40b6097c0ea8d6ac]::run_compiler::{closure#1}>::{closure#0}, core[85c62f5f18d12f5e]::result::Result<(), rustc_span[150ea94e489d3a47]::ErrorGuaranteed>>
  36:     0x7fb5740ae856 - std[88ff7485e6789050]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[6d78e37740b0e5ff]::util::run_in_thread_pool_with_globals<rustc_interface[6d78e37740b0e5ff]::interface::run_compiler<core[85c62f5f18d12f5e]::result::Result<(), rustc_span[150ea94e489d3a47]::ErrorGuaranteed>, rustc_driver_impl[40b6097c0ea8d6ac]::run_compiler::{closure#1}>::{closure#0}, core[85c62f5f18d12f5e]::result::Result<(), rustc_span[150ea94e489d3a47]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[85c62f5f18d12f5e]::result::Result<(), rustc_span[150ea94e489d3a47]::ErrorGuaranteed>>
  37:     0x7fb5740dcd56 - std[88ff7485e6789050]::panicking::try::<core[85c62f5f18d12f5e]::result::Result<(), rustc_span[150ea94e489d3a47]::ErrorGuaranteed>, core[85c62f5f18d12f5e]::panic::unwind_safe::AssertUnwindSafe<<std[88ff7485e6789050]::thread::Builder>::spawn_unchecked_<rustc_interface[6d78e37740b0e5ff]::util::run_in_thread_pool_with_globals<rustc_interface[6d78e37740b0e5ff]::interface::run_compiler<core[85c62f5f18d12f5e]::result::Result<(), rustc_span[150ea94e489d3a47]::ErrorGuaranteed>, rustc_driver_impl[40b6097c0ea8d6ac]::run_compiler::{closure#1}>::{closure#0}, core[85c62f5f18d12f5e]::result::Result<(), rustc_span[150ea94e489d3a47]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[85c62f5f18d12f5e]::result::Result<(), rustc_span[150ea94e489d3a47]::ErrorGuaranteed>>::{closure#1}::{closure#0}>>
  38:     0x7fb57409ccbd - <<std[88ff7485e6789050]::thread::Builder>::spawn_unchecked_<rustc_interface[6d78e37740b0e5ff]::util::run_in_thread_pool_with_globals<rustc_interface[6d78e37740b0e5ff]::interface::run_compiler<core[85c62f5f18d12f5e]::result::Result<(), rustc_span[150ea94e489d3a47]::ErrorGuaranteed>, rustc_driver_impl[40b6097c0ea8d6ac]::run_compiler::{closure#1}>::{closure#0}, core[85c62f5f18d12f5e]::result::Result<(), rustc_span[150ea94e489d3a47]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[85c62f5f18d12f5e]::result::Result<(), rustc_span[150ea94e489d3a47]::ErrorGuaranteed>>::{closure#1} as core[85c62f5f18d12f5e]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  39:     0x7fb5735e945e - std::sys::unix::thread::Thread::new::thread_start::h04412f818bdd9614
  40:     0x7fb573384b43 - <unknown>
  41:     0x7fb573416a00 - <unknown>
  42:                0x0 - <unknown>
note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.71.0-nightly (e12d62e50 2023-05-17) running on x86_64-unknown-linux-gnu


note: compiler flags: -Z threads=1 -C codegen-units=1 -Z ui-testing -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z deduplicate-diagnostics=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0
query stack during panic:
query stack during panic:
#0 [typeck] type-checking `delta`
#1 [used_trait_imports] finding used_trait_imports `delta`
#2 [analysis] running analysis passes on this crate
error: aborting due to 6 previous errors

For more information about this error, try `rustc --explain E0277`.
------------------------------------------
```

</details>

--- a/library/core/tests/mem.rs
+++ b/library/core/tests/mem.rs
@@ -386,6 +386,21 @@ fn offset_of() {
     // Layout of tuples is unstable
     assert!(offset_of!((u8, u16), 0) <= size_of::<(u8, u16)>() - 1);
     assert!(offset_of!((u8, u16), 1) <= size_of::<(u8, u16)>() - 2);
+
+    #[repr(C)]
+    struct Generic<T> {
+        x: u8,
+        y: u32,
+        z: T
+    }
+
+    // Ensure that this type of generics works
+    fn offs_of_z<T>() -> usize {
+        offset_of!(Generic<T>, z)
+    }
+
+    assert_eq!(offset_of!(Generic<u8>, z), 8);
+    assert_eq!(offs_of_z::<u8>(), 8);
 }
 
 #[test]

--- a/tests/ui/offset-of/offset-of-dst-field.rs
+++ b/tests/ui/offset-of/offset-of-dst-field.rs
@@ -26,8 +26,23 @@ struct Gamma {
     z: Extern,
 }
 
+struct Delta<T: ?Sized> {
+    x: u8,
+    y: u16,
+    z: T,
+}
+
 fn main() {
     offset_of!(Alpha, z); //~ ERROR the size for values of type
     offset_of!(Beta, z); //~ ERROR the size for values of type
     offset_of!(Gamma, z); //~ ERROR the size for values of type
+}
+
+fn delta() {
+    offset_of!(Delta<Alpha>, z); //~ ERROR the size for values of type
+    offset_of!(Delta<Extern>, z); //~ ERROR the size for values of type
+}
+
+fn generic_with_maybe_sized<T: ?Sized>() -> usize {
+    offset_of!(Delta<T>, z) //~ ERROR the size for values of type
 }

--- a/tests/ui/offset-of/offset-of-dst-field.stderr
+++ b/tests/ui/offset-of/offset-of-dst-field.stderr
@@ -1,5 +1,5 @@
 error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
-  --> $DIR/offset-of-dst-field.rs:30:5
+  --> $DIR/offset-of-dst-field.rs:36:5
    |
 LL |     offset_of!(Alpha, z);
    |     ^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
@@ -8,7 +8,7 @@ LL |     offset_of!(Alpha, z);
    = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)
 
 error[E0277]: the size for values of type `(dyn Trait + 'static)` cannot be known at compilation time
-  --> $DIR/offset-of-dst-field.rs:31:5
+  --> $DIR/offset-of-dst-field.rs:37:5
    |
 LL |     offset_of!(Beta, z);
    |     ^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
@@ -17,7 +17,7 @@ LL |     offset_of!(Beta, z);
    = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)
 
 error[E0277]: the size for values of type `Extern` cannot be known at compilation time
-  --> $DIR/offset-of-dst-field.rs:32:5
+  --> $DIR/offset-of-dst-field.rs:38:5
    |
 LL |     offset_of!(Gamma, z);
    |     ^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
@@ -25,6 +25,44 @@ LL |     offset_of!(Gamma, z);
    = help: the trait `Sized` is not implemented for `Extern`
    = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)
 
-error: aborting due to 3 previous errors
+error[E0277]: the size for values of type `Extern` cannot be known at compilation time
+  --> $DIR/offset-of-dst-field.rs:43:5
+   |
+LL |     offset_of!(Delta<Extern>, z);
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
+   |
+   = help: the trait `Sized` is not implemented for `Extern`
+   = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
+  --> $DIR/offset-of-dst-field.rs:42:5
+   |
+LL |     offset_of!(Delta<Alpha>, z);
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
+   |
+   = help: within `Alpha`, the trait `Sized` is not implemented for `[u8]`
+note: required because it appears within the type `Alpha`
+  --> $DIR/offset-of-dst-field.rs:5:8
+   |
+LL | struct Alpha {
+   |        ^^^^^
+   = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error[E0277]: the size for values of type `T` cannot be known at compilation time
+  --> $DIR/offset-of-dst-field.rs:47:5
+   |
+LL | fn generic_with_maybe_sized<T: ?Sized>() -> usize {
+   |                             - this type parameter needs to be `std::marker::Sized`
+LL |     offset_of!(Delta<T>, z)
+   |     ^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
+   |
+   = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)
+help: consider removing the `?Sized` bound to make the type parameter `Sized`
+   |
+LL - fn generic_with_maybe_sized<T: ?Sized>() -> usize {
+LL + fn generic_with_maybe_sized<T>() -> usize {
+   |
+
+error: aborting due to 6 previous errors
 
 For more information about this error, try `rustc --explain E0277`.

--- a/tests/ui/offset-of/offset-of-output-type.rs
+++ b/tests/ui/offset-of/offset-of-output-type.rs
@@ -1,0 +1,20 @@
+#![feature(offset_of)]
+
+use std::mem::offset_of;
+
+struct S {
+    v: u8,
+    w: u16,
+}
+
+
+fn main() {
+    let _: u8 = offset_of!(S, v); //~ ERROR mismatched types
+    let _: u16 = offset_of!(S, v); //~ ERROR mismatched types
+    let _: u32 = offset_of!(S, v); //~ ERROR mismatched types
+    let _: u64 = offset_of!(S, v); //~ ERROR mismatched types
+    let _: isize = offset_of!(S, v); //~ ERROR mismatched types
+    let _: usize = offset_of!(S, v);
+
+    offset_of!(S, v) //~ ERROR mismatched types
+}

--- a/tests/ui/offset-of/offset-of-output-type.stderr
+++ b/tests/ui/offset-of/offset-of-output-type.stderr
@@ -1,0 +1,64 @@
+error[E0308]: mismatched types
+  --> $DIR/offset-of-output-type.rs:12:17
+   |
+LL |     let _: u8 = offset_of!(S, v);
+   |            --   ^^^^^^^^^^^^^^^^ expected `u8`, found `usize`
+   |            |
+   |            expected due to this
+   |
+   = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error[E0308]: mismatched types
+  --> $DIR/offset-of-output-type.rs:13:18
+   |
+LL |     let _: u16 = offset_of!(S, v);
+   |            ---   ^^^^^^^^^^^^^^^^ expected `u16`, found `usize`
+   |            |
+   |            expected due to this
+   |
+   = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error[E0308]: mismatched types
+  --> $DIR/offset-of-output-type.rs:14:18
+   |
+LL |     let _: u32 = offset_of!(S, v);
+   |            ---   ^^^^^^^^^^^^^^^^ expected `u32`, found `usize`
+   |            |
+   |            expected due to this
+   |
+   = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error[E0308]: mismatched types
+  --> $DIR/offset-of-output-type.rs:15:18
+   |
+LL |     let _: u64 = offset_of!(S, v);
+   |            ---   ^^^^^^^^^^^^^^^^ expected `u64`, found `usize`
+   |            |
+   |            expected due to this
+   |
+   = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error[E0308]: mismatched types
+  --> $DIR/offset-of-output-type.rs:16:20
+   |
+LL |     let _: isize = offset_of!(S, v);
+   |            -----   ^^^^^^^^^^^^^^^^ expected `isize`, found `usize`
+   |            |
+   |            expected due to this
+   |
+   = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error[E0308]: mismatched types
+  --> $DIR/offset-of-output-type.rs:19:5
+   |
+LL | fn main() {
+   |           - expected `()` because of default return type
+...
+LL |     offset_of!(S, v)
+   |     ^^^^^^^^^^^^^^^^ expected `()`, found `usize`
+   |
+   = note: this error originates in the macro `offset_of` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error: aborting due to 6 previous errors
+
+For more information about this error, try `rustc --explain E0308`.

--- a/tests/ui/offset-of/offset-of-private.rs
+++ b/tests/ui/offset-of/offset-of-private.rs
@@ -8,9 +8,21 @@ mod m {
         pub public: u8,
         private: u8,
     }
+    #[repr(C)]
+    pub struct FooTuple(pub u8, u8);
+    #[repr(C)]
+    struct Bar {
+        pub public: u8,
+        private: u8,
+    }
 }
 
 fn main() {
     offset_of!(m::Foo, public);
     offset_of!(m::Foo, private); //~ ERROR field `private` of struct `Foo` is private
+    offset_of!(m::FooTuple, 0);
+    offset_of!(m::FooTuple, 1); //~ ERROR field `1` of struct `FooTuple` is private
+    offset_of!(m::Bar, public); //~ ERROR struct `Bar` is private
+    offset_of!(m::Bar, private); //~ ERROR struct `Bar` is private
+    //~| ERROR field `private` of struct `Bar` is private
 }

--- a/tests/ui/offset-of/offset-of-private.stderr
+++ b/tests/ui/offset-of/offset-of-private.stderr
@@ -1,9 +1,46 @@
+error[E0603]: struct `Bar` is private
+  --> $DIR/offset-of-private.rs:25:19
+   |
+LL |     offset_of!(m::Bar, public);
+   |                   ^^^ private struct
+   |
+note: the struct `Bar` is defined here
+  --> $DIR/offset-of-private.rs:14:5
+   |
+LL |     struct Bar {
+   |     ^^^^^^^^^^
+
+error[E0603]: struct `Bar` is private
+  --> $DIR/offset-of-private.rs:26:19
+   |
+LL |     offset_of!(m::Bar, private);
+   |                   ^^^ private struct
+   |
+note: the struct `Bar` is defined here
+  --> $DIR/offset-of-private.rs:14:5
+   |
+LL |     struct Bar {
+   |     ^^^^^^^^^^
+
 error[E0616]: field `private` of struct `Foo` is private
-  --> $DIR/offset-of-private.rs:15:24
+  --> $DIR/offset-of-private.rs:22:24
    |
 LL |     offset_of!(m::Foo, private);
    |                        ^^^^^^^ private field
 
-error: aborting due to previous error
+error[E0616]: field `1` of struct `FooTuple` is private
+  --> $DIR/offset-of-private.rs:24:29
+   |
+LL |     offset_of!(m::FooTuple, 1);
+   |                             ^ private field
 
-For more information about this error, try `rustc --explain E0616`.
+error[E0616]: field `private` of struct `Bar` is private
+  --> $DIR/offset-of-private.rs:26:24
+   |
+LL |     offset_of!(m::Bar, private);
+   |                        ^^^^^^^ private field
+
+error: aborting due to 5 previous errors
+
+Some errors have detailed explanations: E0603, E0616.
+For more information about an error, try `rustc --explain E0603`.

--- a/tests/ui/offset-of/offset-of-self.rs
+++ b/tests/ui/offset-of/offset-of-self.rs
@@ -1,0 +1,58 @@
+#![feature(offset_of)]
+
+use std::mem::offset_of;
+
+struct C<T> {
+    v: T,
+    w: T,
+}
+
+struct S {
+    v: u8,
+    w: u16,
+}
+
+impl S {
+    fn v_offs() -> usize {
+        offset_of!(Self, v)
+    }
+    fn v_offs_wrong_syntax() {
+        offset_of!(Self, Self::v); //~ ERROR no rules expected the token `::`
+        offset_of!(S, Self); //~ ERROR expected identifier, found keyword `Self`
+        //~| no field `Self` on type `S`
+    }
+    fn offs_in_c() -> usize {
+        offset_of!(C<Self>, w)
+    }
+    fn offs_in_c_colon() -> usize {
+        offset_of!(C::<Self>, w)
+    }
+}
+
+mod m {
+    use std::mem::offset_of;
+    fn off() {
+        offset_of!(self::S, v); //~ ERROR cannot find type `S` in module
+        offset_of!(super::S, v);
+        offset_of!(crate::S, v);
+    }
+    impl super::n::T {
+        fn v_offs_self() -> usize {
+            offset_of!(Self, v) //~ ERROR field `v` of struct `T` is private
+        }
+    }
+}
+
+mod n {
+    pub struct T { v: u8, }
+}
+
+fn main() {
+    offset_of!(self::S, v);
+    offset_of!(Self, v); //~ ERROR cannot find type `Self` in this scope
+
+    offset_of!(S, self); //~ ERROR expected identifier, found keyword `self`
+    //~| no field `self` on type `S`
+    offset_of!(S, v.self); //~ ERROR expected identifier, found keyword `self`
+    //~| no field `self` on type `u8`
+}

--- a/tests/ui/offset-of/offset-of-self.stderr
+++ b/tests/ui/offset-of/offset-of-self.stderr
@@ -1,0 +1,79 @@
+error: no rules expected the token `::`
+  --> $DIR/offset-of-self.rs:20:30
+   |
+LL |         offset_of!(Self, Self::v);
+   |                              ^^ no rules expected this token in macro call
+   |
+   = note: while trying to match sequence start
+
+error: expected identifier, found keyword `Self`
+  --> $DIR/offset-of-self.rs:21:23
+   |
+LL |         offset_of!(S, Self);
+   |                       ^^^^ expected identifier, found keyword
+
+error: expected identifier, found keyword `self`
+  --> $DIR/offset-of-self.rs:54:19
+   |
+LL |     offset_of!(S, self);
+   |                   ^^^^ expected identifier, found keyword
+
+error: expected identifier, found keyword `self`
+  --> $DIR/offset-of-self.rs:56:21
+   |
+LL |     offset_of!(S, v.self);
+   |                     ^^^^ expected identifier, found keyword
+
+error[E0412]: cannot find type `S` in module `self`
+  --> $DIR/offset-of-self.rs:35:26
+   |
+LL |         offset_of!(self::S, v);
+   |                          ^ not found in `self`
+   |
+help: consider importing this struct
+   |
+LL +     use S;
+   |
+help: if you import `S`, refer to it directly
+   |
+LL -         offset_of!(self::S, v);
+LL +         offset_of!(S, v);
+   |
+
+error[E0411]: cannot find type `Self` in this scope
+  --> $DIR/offset-of-self.rs:52:16
+   |
+LL | fn main() {
+   |    ---- `Self` not allowed in a function
+LL |     offset_of!(self::S, v);
+LL |     offset_of!(Self, v);
+   |                ^^^^ `Self` is only available in impls, traits, and type definitions
+
+error[E0609]: no field `Self` on type `S`
+  --> $DIR/offset-of-self.rs:21:23
+   |
+LL |         offset_of!(S, Self);
+   |                       ^^^^
+
+error[E0616]: field `v` of struct `T` is private
+  --> $DIR/offset-of-self.rs:41:30
+   |
+LL |             offset_of!(Self, v)
+   |                              ^ private field
+
+error[E0609]: no field `self` on type `S`
+  --> $DIR/offset-of-self.rs:54:19
+   |
+LL |     offset_of!(S, self);
+   |                   ^^^^
+
+error[E0609]: no field `self` on type `u8`
+  --> $DIR/offset-of-self.rs:56:21
+   |
+LL |     offset_of!(S, v.self);
+   |                     ^^^^
+
+error: aborting due to 10 previous errors
+
+Some errors have detailed explanations: E0411, E0412, E0609, E0616.
+For more information about an error, try `rustc --explain E0411`.

